### ISSUE TITLE
Is Read required on top of Write for PATCH delete and where ?

### ISSUE THREAD

**Issue by bourgeoa** [solid/specification#220]
On PATCH spec seems to require only `Write` for sparql DELETE. This is implemented in solid/test-suite.
But NSS is requiring `Read and Write`, with the following explanations in the code comments in NSS PATCH implementation.

```
// Read access is required for DELETE and WHERE.
// If we would allows users without read access,
// they could use DELETE or WHERE to trigger 200 or 409,
// and thereby guess the existence of certain triples.
// DELETE additionally requires write access.
```
What should be the spec ? What should test-suite implement if the response is not straight forward ?

**Comment by RubenVerborgh** [comment]
I've come up with the above explanation together with @timbl, and stand by it.

**Comment by csarven** [comment]
I'm not sure which spec you're referring to but will take it as underspecified/unknown at the very least in the issues here and current Protocol spec. But right that the current web-access-control-spec alludes to `acl:Write` allowing PATCH.. and the wiki WebAccessControl doesn't require Read.

---

When using WAC, I agree on requiring `acl:Read` and `acl:Write` in order to accept PATCH operations including `DELETE DATA` or `WHERE` - at least those with application/sparql-update data type if anything.

When agent doesn't have `acl:Read` and `acl:Write` on target, rejection should use the 403 status code.

When agent has `acl:Read` and `acl:Write` on target and there is *no match* on `DELETE DATA`, rejection should use the 409 status code.

I've documented some PATCH requests in a table based on `INSERT DATA` and `DELETE DATA` in https://github.com/solid/specification/issues/14#issuecomment-683480525 as those appear to be the basic level of SPARQL Update we seem to agree on to date. Related issue: https://github.com/solid/specification/issues/125 . So, `WHERE` support needs an explicit agreement.

See also for recent-ish description of the current state of things: https://github.com/solid/community-server/issues/15#issuecomment-668615622

**Comment by RubenVerborgh** [comment]
Note that this also relates to #139. If SPARQL UPDATE does _not_ act as a semaphore mechanism (so if it _does_ follow the SPARQL 1.1 spec), then we might not be leaking any information.

So perhaps WebAccessControl was initially assuming that semantics; and its only when the semaphore semantics were introduced that the leakage was created.

**Comment by RubenVerborgh** [comment]
@acoburn What does ESS implement?

**Comment by acoburn** [comment]
With ESS, a PATCH operation does not _require_ READ access. 

There is also no locking mechanism and hence no notion of a semaphore in ESS (distributed locks are a nightmare); instead, it relies on a conflict resolution mechanism (e.g. last-write-wins)

**Comment by acoburn** [comment]
Worth noting that one could apply a similar line of reasoning about conditional PUT requests (`If-Match`, etc)

**Comment by RubenVerborgh** [comment]
> With ESS, a PATCH operation does not _require_ READ access.

Check; so this might substantiate my theory that the Read access is an artifact of the non-standard SPARQL semantics.

Because with standard SPARQL, they would all be 2xx responses, even if some triples were not present for deletion, or some inserts were not successful because they already existed.

> There is also no locking mechanism and hence no notion of a semaphore in ESS (distributed locks are a nightmare); instead, it relies on a conflict resolution mechanism (e.g. last-write-wins)

Just a small note here for everyone's understanding: there's two kinds of scale here:
1. two/multiple requests happening roughly at the same time
2. two request happening some time apart, while a third party has changed the resource in between

I _think_ Aaron might be talking only or mainly about case 1.

In contrast, @timbl was mainly thinking about case 2, where 2 agents should be able to modify different parts of the same document independently (but errors if those edits overlap). So a more coarse granularity than `If-Match`.

**Comment by michielbdejong** [comment]
So now that the test suite [was updated 20 days ago](https://github.com/solid/web-access-control-tests/commit/d26e566a9d3faedd8e8c3167052fe105a67b7e98) to require Read access, [CSS is failing that test](https://github.com/solid/web-access-control-tests/blob/a90373499b9ebdb15f982ffa995797c299975ef5/test/surface/update.test.ts#L575-L577). Marking that test as disputed now.

We probably can't answer this question without answering https://github.com/solid/specification/issues/139 first.

**Comment by kjetilk** [comment]
OK, catching up here... 

Indeed, the `Read` requirement stems from the semaphore mechanism, because that leaks information. 

The deeper problem here is really how we design systems that allow for actual least privileges to be used. It is easy to say that it should follow the principle of least privilege, a different thing to design for it. To design for it, we need to make sure that all of SPARQL semantics maps well to access modes. 

We could take the easy way out and have the semaphore mechanism and say that all SPARQL Update queries require Read+Write, but that wouldn't be a great design in light of the principle of least privilege. Then, we could also ensure ACID in the database but not allow these semaphores, but that puts the ACID requirement on all implementations, which is also a lot to ask. We could also spec a lot of special cases around it. 

I gave this some thought a long time ago, and I think there is a reasonable middle ground, [but it requires changes to SPARQL](https://lists.w3.org/Archives/Public/public-sparql-12/2020Jan/0000.html).

**Comment by michielbdejong** [comment]
Conclusion: clients should just avoid ever trying to delete a triple that's not present, because as we've seen, different pod servers will react differently. Instead, a PATCH that deletes triples should always have an `If-Match` header that avoids race conditions.

**Comment by kjetilk** [comment]
This has been discussed a bit amongst the Solid Editors, and it has been decided that the semaphore mechanism is a requirement. It is considered that we should come up with a better design the future, but for now, that is the case. That decision makes it clear that the answer to the title of this issue is *yes*.

AFAIK, the mechanism currently applies to the `WHERE` clause, but I think it is quite clear that the use case would naturally extend to a request with `INSERT DATA` and `DELETE DATA` as well. I also think timescale is a poor judge. In [the post to SPARQL 1.2 CG](https://lists.w3.org/Archives/Public/public-sparql-12/2020Jan/0000.html), I use an example that involves these two operations, like this:

Concretely, say that client 1 goes:

```sparql
DELETE DATA { <foo> <baz> "Dahut" } ;
INSERT DATA { <foo> <baz> "Bar" }
```

independently, client 2 goes 

```sparql
DELETE DATA { <foo> <baz> "Dahut" } ;
INSERT DATA { <foo> <baz> "Foobar" }
```

before the first client as finished. The same semaphore would apply, as the server would respond with a `409` to the `DELETE DATA` operation of the second client.

I realize that it would be a simpler implementation if the semaphore mechanism applied only to a `WHERE` clause, but it seems rather odd to tie it to that, given the above.

Thus, it seems to me that for now, we should require Read+Write privs for any SPARQL query with `DELETE`. 

I hope that we can get to a SPARQL standard where any information exposure is tied to a projection, so that Read would only apply if you actually use the mechanism, but for now, I think the above is the best way to do it.

**Comment by kjetilk** [comment]
Can we arrive at a consensus that the semaphore mechanism applies to both `DELETE` operations, so that Read are required for `PATCH` containing those?

**Comment by csarven** [comment]
+0.5.

Just want to highlight that the work in https://github.com/solid/specification/issues/14#issuecomment-683480525 is generally based on:

>Resources can be observable or discoverable - "knowable" - by agents having Read access privilege either on the resource or its container (inherited).

While it is possible to distinguish between knowledge about the existence of a resource from the description of a resource, the general principle, i.e., requiring Read, seems to apply to both.

**Comment by kjetilk** [comment]
The Solid Editors Meeting today, with (just) @csarven and @kjetilk present resolved:

The semaphore mechanism applies to all `DELETE` queries, and therefore read and write are required.

**Comment by RubenVerborgh** [comment]
> The semaphore mechanism applies to all `DELETE` queries, and therefore read and write are required.

Does that mean that https://github.com/solid/specification/issues/139 is also resolved?

**Comment by kjetilk** [comment]
Yes, that will be resolved when we close #125

**Comment by kjetilk** [comment]
This issue has been nominated for drafting phase for the next milestone.
